Subbus/Classes/Subbus.swift

```diff
@@ -8,141 +8,241 @@
-    }
-    
-    //Misc
-    static func log(_ message: String) {
-        guard loggingEnabled == true else { return }
-        print("Subbus: \(message)")
+    public static func unsubscribe<T>(event: T.Type) {
+        let eventType = String(reflecting: T.self)
+        subscriptions.removeAll(where: { $0.matches(eventType: eventType) })
     }
-    
-    static func stringFor<I>(id: I) -> String? {
+}
+
+extension Subbus {
+    internal static func stringFor(id: Any) -> String? {
         var valString = "" //Value or pointer
-        
+        let idType = type(of: id)
+
         //Get the right string type; objects need memory address and primitives just need value.
-        if I.self is AnyClass {
+        if idType is AnyClass {
             valString = "\(Unmanaged.passUnretained(id as AnyObject).toOpaque())"
         } else {
             valString = "\(id)"
         }
-        
+
         //Ensure that the idString is valid
         guard valString.isEmpty == false else { return nil }
         guard valString != "nil" else { return nil }
+
+        return "\(idType)-\(valString)"
+    }
+    
+    internal static func log(_ message: String, force: Bool = false) {
+        guard logToConsole || force else { return }
+        print("Subbus: \(message)")
+    }
+    
+    internal static func cleanup() {
+        subscriptions.removeAll(where: { $0.identifier == nil })
+    }
+    
+    internal static func clearHistory() {
+        unhandledEvents = []
+        log("WARNING: cleared event history!  This function was made for unit testing, please don't use it.", force: true)
+    }
+    
+    internal static func clearSubscribers() {
+        subscriptions = []
+        log("WARNING: cleared subscriber cache!  This function was made for unit testing, please don't use it.", force: true)
+    }
+}
+
+internal struct Subscription {
+    weak var identifier: AnyObject?
+    var identifierString: String = ""
+    var eventType: String
+    var handler: Any
+    
+    init?(identifier: Any, eventType: String, handler: Any) {
+        self.eventType = eventType
+        self.handler = handler
+        
+        let (id, message) = parseId(identifier)
+        
+        guard let parsedId = id, message == nil else {
+            Subbus.log(message ?? "Subbus: Unable to recognize identifier", force: true)
+            return nil
+        }
+        
+        self.identifier = parsedId
         
-        let name = "\(I.self)-\(valString)"
-        return name
+        if let stringId = parsedId as? String {
+            self.identifierString = stringId
+        } else {
+            self.identifierString = Subbus.stringFor(id: parsedId) ?? String(describing: parsedId)
+        }
+    }
+    
+    private func parseId(_ id: Any) -> (AnyObject?, String?) {
+        let mirror = Mirror(reflecting: id)
+        
+        switch mirror.displayStyle {
```

Comment:
This is very interesting; I didn't know we could do this. I'm going to read into this more.

Reply:
It's awesome!  It's frankly the *core* of this refactor.  Optionals are unwrapped, classes are held with a weak reference, and everything else uses the "old" make-a-string-to-represent-it method.